### PR TITLE
[mlir][Vector] Add support for trunci to narrow type emulation

### PR DESCRIPTION
**Pull request by dcaballe**
This PR add support for `arith.trunci` to vector narrow type emulation for iX -> i4 truncations, for X >= 8. For now, the pattern only works for 1D vectors and is based on `vector.shuffle` ops. We would need `vector.deinterleave` to add n-D vector support.

### CODE REVIEW

**Comment by dcaballe** [comment]
Kind ping

File: mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp [hunk 1]
```diff
@@ -876,6 +876,57 @@ static Value rewriteI4ToI8SignedExt(PatternRewriter &rewriter, Location loc,
...
+/// bitwise ops that take advantage of high-level information to avoid leaving
+/// LLVM to scramble with peephole optimizations.
+static Value rewriteI8ToI4Trunc(PatternRewriter &rewriter, Location loc,
+                                Value srcValue) {
+  VectorType srcVecType = cast<VectorType>(srcValue.getType());
+  assert(srcVecType.getElementType().isSignlessInteger(8) &&
+         "Expected i8 type");
+
+  // 1. De-interleave low and high i8 elements.
+  int64_t vecDimSize = srcVecType.getShape().back();
```
Comment (MacDue) [inline mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:889]: Could `vecDimSize` not be divisible by 2?

**Review by MacDue — APPROVED** [review]
LGTM, just a some minor comments.

File: mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp (hunk 1 above)
Comment (dcaballe) [reply mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:889]: Odd number of i4 elements is not supported. I added that to the preconditions and an assert here.

**Pull request merged**